src/wp-includes/blocks/navigation.php

```diff
@@ -1066,3 +1342,120 @@ function block_core_navigation_get_most_recently_published_navigation() {
 
 	return null;
 }
+
+/**
+ * Insert hooked blocks into a Navigation block.
+ *
+ * Given a Navigation block's inner blocks and its corresponding `wp_navigation` post object,
+ * this function inserts hooked blocks into it, and returns the serialized inner blocks in a
+ * mock Navigation block wrapper.
+ *
+ * If there are any hooked blocks that need to be inserted as the Navigation block's first or last
+ * children, the `wp_navigation` post's `_wp_ignored_hooked_blocks` meta is checked to see if any
+ * of those hooked blocks should be exempted from insertion.
+ *
+ * @param array   $inner_blocks Parsed inner blocks of a Navigation block.
+ * @param WP_Post $post         `wp_navigation` post object corresponding to the block.
+ * @return string Serialized inner blocks in mock Navigation block wrapper, with hooked blocks inserted, if any.
+ */
+function block_core_navigation_insert_hooked_blocks( $inner_blocks, $post = null ) {
+	$before_block_visitor = null;
+	$after_block_visitor  = null;
+	$hooked_blocks        = get_hooked_blocks();
+	$attributes           = array();
+
+	if ( isset( $post->ID ) ) {
+		$ignored_hooked_blocks = get_post_meta( $post->ID, '_wp_ignored_hooked_blocks', true );
+		if ( ! empty( $ignored_hooked_blocks ) ) {
+			$ignored_hooked_blocks  = json_decode( $ignored_hooked_blocks, true );
+			$attributes['metadata'] = array(
+				'ignoredHookedBlocks' => $ignored_hooked_blocks,
+			);
+		}
+	}
+
+	$mock_anchor_parent_block = array(
+		'blockName'    => 'core/navigation',
+		'attrs'        => $attributes,
+		'innerBlocks'  => $inner_blocks,
+		'innerContent' => array_fill( 0, count( $inner_blocks ), null ),
+	);
+	$before_block_visitor     = null;
+	$after_block_visitor      = null;
+
+	if ( ! empty( $hooked_blocks ) || has_filter( 'hooked_block_types' ) ) {
+		$before_block_visitor = make_before_block_visitor( $hooked_blocks, $post );
+		$after_block_visitor  = make_after_block_visitor( $hooked_blocks, $post );
+	}
+
+	return traverse_and_serialize_block( $mock_anchor_parent_block, $before_block_visitor, $after_block_visitor );
+}
+
+/**
+ * Updates the post meta with the list of ignored hooked blocks when the navigation is created or updated via the REST API.
+ *
+ * @param WP_Post $post Post object.
+ */
+function block_core_navigation_update_ignore_hooked_blocks_meta( $post ) {
+	if ( ! isset( $post->ID ) ) {
```

Comment:
Why would the post ID not exist, and why is a silent return appropriate rather than and error/warning/notice?

Reply:
cc @tjcafferkey

Reply:
Checking for the existence of the post ID was merely a safety check, I guess it’s a habit for me to never assume data is always going to exist. That said, the API pretty much guarantees `$post` is always going to exist at this point so I've removed it in https://github.com/WordPress/gutenberg/pull/58379